notification_query.go

```diff
@@ -0,0 +1,248 @@
+// Should be called before query being started.
+//
+// Args reference: https://docs.microsoft.com/en-us/windows/desktop/wmisdk/swbemlocator-connectserver
+// Passing details: https://github.com/go-ole/go-ole/blob/master/idispatch_windows.go#L60
+func (q *NotificationQuery) SetConnectServerArgs(args ...interface{}) {
+	q.connectServerArgs = args
+}
+
+// StartNotifications connects to the WMI service and starts receiving
+// notifications generated by the query.
+//
+// Errors are usually happen on initialization phase (connect to WMI,
+// query execution, first result unmarshalling) so you could assume that
+// "it's either starts and going to give me notifications or fails fast enough".
+func (q *NotificationQuery) StartNotifications() (err error) {
+	q.Lock()
+	if q.isRunning {
+		q.Unlock()
+		return ErrAlreadyRunning
+	}
+	q.isRunning = true
+	q.Unlock()
+
+	//  Be aware of reflections and COM usage.
+	defer func() {
+		if r := recover(); r != nil {
+			err = multierror.Append(err, fmt.Errorf("runtime panic; %v", err))
+		}
+	}()
+
+	// Notify that we are going to use COM.
+	comshim.Add(1)
+	defer comshim.Done()
+
+	// Connect to WMI service.
+	service, err := createWMIConnection(q.connectServerArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to connect WMI service; %s", err)
+	}
+	defer service.Release()
+
+	// Subscribe to the events. ExecNotificationQuery call must have that flags
+	// and no other.
+	sWbemEventSource, err := oleutil.CallMethod(
+		service,
+		"ExecNotificationQuery",
+		q.query,
+		"WQL",
+		0x00000010|0x00000020, // WBEM_FLAG_RETURN_IMMEDIATELY | WBEM_FLAG_FORWARD_ONLY
+	)
+	if err != nil {
+		return fmt.Errorf("ExecNotificationQuery failed; %s", err)
+	}
+	eventSource := sWbemEventSource.ToIDispatch()
+	defer eventSource.Release()
+
+	reflectedDoneChan := reflect.ValueOf(q.doneCh)
+	reflectedResChan := reflect.ValueOf(q.eventCh)
+	eventType := reflectedResChan.Type().Elem()
+	for {
+		// If it is a time to stop somebody will listen on doneCh.
+		select {
+		case q.doneCh <- struct{}{}:
+			return nil
+		default:
+		}
+
+		// Or try to query new events waiting no longer than queryTimeoutMs.
+		eventIUknown, err := eventSource.CallMethod("NextEvent", q.queryTimeoutMs)
+		if err != nil {
+			if isTimeoutError(err) {
+				continue
+			}
+			return fmt.Errorf("unexpected NextEvent error; %s", err)
+		}
+		event := eventIUknown.ToIDispatch()
+
+		// Unmarshal event.
+		e := reflect.New(eventType)
+		if err := q.Unmarshal(event, e.Interface()); err != nil {
+			return fmt.Errorf("failed to unmarshal event; %s", err)
+		}
+
+		// Send to the user.
+		sent := trySend(reflectedResChan, reflectedDoneChan, e.Elem())
+		if !sent {
+			return nil // Query stopped
+		}
+	}
+}
+
+// Stop stops the running query waiting until everything is released. It could
+// take some time for query to receive a stop signal. See `SetNotificationTimeout`
+// for more info.
+func (q *NotificationQuery) Stop() {
```

Comment:
Not sure if it's not intended behaviour but: if you call `Stop()` prior to receive from `eventCh`, `StartNotifications()` will never return. It worth to document or eliminate the condition.

Reply:
Nice catch!

The problem was not in `Stop()` before receiving, but in `Stop()` before the actual `StartNotifications()`.

Make the code more foolproof now.